### ISSUE TITLE
TTV.LOL leaking userID with IP

### ISSUE THREAD

**Issue by saucettv** [pixeltris/TwitchAdSolutions#32]
As posted on Reddit:

"Just a warning to those using TTV LOL, especially if you are a streamer. The extension currently leaks your Twitch user ID and personal IP to their Russian proxy. For obvious reasons this could be a problem. I have pasted an example below of the redirect URL used.

api.ttv.lol/playlist/streamer.m3u8%3Fallow_source%3Dtrue%26fast_bread%3Dtrue%26p%3A2255857%26play_session_id%3Ad453f480gg1r99h45a3095366ca5363a%26player_backend%3Dmediaplayer%26playlist_include_framerate%3Dtrue%26reassignments_supported%3Dtrue%26sig%3D86df4f12ab1d2ba8d99f82a81680be74e549c67a%26supported_codecs%3Dvp09%252Cavc1%26token%3D%257B%2522adblock%2522%253Afalse%252C%2522authorization%2522%253A%257B%2522forbidden%2522%253Afalse%252C%2522reason%2522%253A%2522%2522%257D%252C%2522blackout_enabled%2522%253Afalse%252C%2522channel%2522%253A%2522streamer%2522%252C%2522channel_id%2522%253A44338537%252C%2522chansub%2522%253A%257B%2522restricted_bitrates%2522%253A%255B%255D%252C%2522view_until%2522%253A1924905600%257D%252C%2522ci_gb%2522%253Afalse%252C%2522geoblock_reason%2522%253A%2522%2522%252C%2522device_id%2522%253A%2522Ob6oHtGYHjIKKJjjjRDwMe%2522%252C%2522expires%2522%253A1620467116%252C%2522extended_history_allowed%2522%253Afalse%252C%2522game%2522%253A%2522%2522%252C%2522hide_ads%2522%253Afalse%252C%2522https_required%2522%253Atrue%252C%2522mature%2522%253Afalse%252C%2522partner%2522%253Afalse%252C%2522platform%2522%253A%2522web%2522%252C%2522player_type%2522%253A%2522site%2522%252C%2522private%2522%253A%257B%2522allowed_to_view%2522%253Atrue%257D%252C%2522privileged%2522%253Afalse%252C%2522role%2522%253A%2522%2522%252C%2522server_ads%2522%253Atrue%252C%2522show_ads%2522%253Atrue%252C%2522subscriber%2522%253Afalse%252C%2522turbo%2522%253Afalse%252C%2522user_id%2522%253D7665776918%252C%2522user_ip%2522%253A%2522142.36.78.245%2522%252C%2522version%2522%253A2%257D%26cdm%3Dwv%26player_version%3D1.4.0

Hopefully the devs will remove this information from it's extension, as it shouldn't be needed. Not trying to scare people away as I'm sure it was just easier for them to implement it this way but still, it is a security risk.

This does not apply to Purple Ad-Block, which only sends the channel name."

Tested this myself and it does leak that information. Haven't tested purple ad block though.

**Comment by pixeltris** [comment]
It would make sense to make an issue here https://github.com/TTV-LOL/extensions/issues

**Comment by saucettv** [comment]
Added the issue to their repo, it might be worth adding some text that states user id's(usernames) will be linked with personal IP's when requests go to their server.

**Comment by captainswain** [comment]
Posted this comment on the issue on our repo, but will paste it here as this is literally the same issue.

Hi,

I just want to clear up any confusion on this topic. As per our privacy policy, we are not storing your twitch user ids or IP from the request. The endpoint you are hitting is also not some random proxy in Russia. Depending on where you are located you will be hitting a server in new york or Germany where your request is parsed and then proxied. Once the handler call is complete your request falls out of scope and in turn, any identifying information in memory is gone. Once again this data is not stored and is not sent as part of the proxied request. 

It was implemented this way for a couple of reasons:

1. easier to just pull all query parameters
2. We pass your non-identifying arguments as part of the proxied request ie: fast_bread, supported_codecs, etc.


I personally don't have time to refactor the server and plugin currently, but I can put that on our roadmap as a feature. I can see how seeing your User ID is scary, but rest assured we are not storing it or passing it along in our proxied request.

**Comment by pixeltris** [comment]
I'm going to close this. As mentioned there's an issue for this on their repo which you can subscribe to https://github.com/TTV-LOL/extensions/issues/8

I have no real reason to suspect any foul play, and the postings by @captainswain on here and reddit seem genuine in their intent.

I have no plans to add any warning assuming there's a plan to omit the `user_id` parameter at some point.

**Comment by saucettv** [comment]
Although I agree that the dev seems genuine. It is still a security risk that can be avoided fairly easily and not putting a small warning until the issue is fixed, in my opinion is a mistake. I have given a faster option for the dev on that issue, so that they would only need to refactor the extension code, rather than the server code. As we both know, it's something that should really only take a few minutes, so I'm not sure what the delay is exactly, but it does worry me.

**Comment by ChoosenEye** [comment]
if he already changed the Extension Chrome Webstore will take some Days until its get rolled out, but dunno if they changed the code.

imho the full String is not that bad, you really can't extract informations from it, the IP is already known because of the request to their Servers, and the UserID can't be reverse because Twitch has no API Endpoint for it.

**Comment by saucettv** [comment]
Yeah hopefully it is already in review, but like he said, he doesn't have the time right now, so not sure. 

You can convert userID to userName with the public Twitch API Endpoint https://dev.twitch.tv/docs/v5/reference/users#get-user-by-id

The main risk is for people with business internet, that is linked to a home address. Mainly that only applies to large streamers, but could also apply to anyone. So although the risk maybe small, it's still a risk that can easily be solved.

**Comment by ChoosenEye** [comment]
EDIT to me:

Twitch changed this to the original User ID from Twitch, maybe in the Change to the GQL Endpoint for Playlist Generation dunno, okay yeah you could technically reverse which User is Viewing the Stream. Including its personal IP.

So yeah this should be fixed.
